### PR TITLE
feat(matter): initial commit with arduino matter lib

### PR DESCRIPTION
**Pull request by SuGlider**
## Description of Change

Implementation of Matter Library for ESP32 Arduino Core.
Matter is implemented over WiFi.
Commissioning is done using the Pairing Code or the QRCode from an URL.

Initial Device implmented is a on/off Light.
The Library allow adding more than one endpoint to a matter node.

## Tests scenarios
ESP32 / ESP32-S2 / ESP32-S3 / ESP32-C3 / ESP32-C6

Testing done using Alexa Echo Dot and the examples from the Matter Library.

## Related links

Related to #7432

### CODE REVIEW

File: libraries/Matter/examples/MatterCommissionTest/MatterCommissionTest.ino [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+// Matter Manager
+#include <Matter.h>
+
+// List of Matter Endpoints for this Node
+// On/Off Light Endpoint
+#include <MatterOnOffLight.h>
+MatterOnOffLight OnOffLight;
+
+// WiFi is manually set and stated
```
Comment (SuGlider) [inline libraries/Matter/examples/MatterCommissionTest/MatterCommissionTest.ino:14]:
```suggestion
// WiFi is manually set and started

```

File: libraries/Matter/examples/MatterComposedLights/MatterComposedLights.ino [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+  Serial.printf("CB-Light2 changed state to: %s\r\n", state ? "ON" : "OFF");
+  return true;
+}
+
+bool setLightOnOff3(bool state) {
+  Serial.printf("CB-Light3 changed state to: %s\r\n", state ? "ON" : "OFF");
+  return true;
+}
+
+// WiFi is manually set and stated
```
Comment (SuGlider) [inline libraries/Matter/examples/MatterComposedLights/MatterComposedLights.ino:32]:
```suggestion
// WiFi is manually set and started

```

File: libraries/Matter/examples/MatterOnOffLight/MatterOnOffLight.ino [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+  if (state) {
+    digitalWrite(ledPin, HIGH);
+  } else {
+    digitalWrite(ledPin, LOW);
+  }
+  // This callback must return the success state to Matter core
+  return true;
+}
+
+// WiFi is manually set and stated
```
Comment (SuGlider) [inline libraries/Matter/examples/MatterOnOffLight/MatterOnOffLight.ino:34]:
```suggestion
// WiFi is manually set and started

```

File: libraries/Matter/examples/MatterCommissionTest/MatterCommissionTest.ino [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+  Serial.println("\r\nWiFi connected");
+  Serial.println("IP address: ");
+  Serial.println(WiFi.localIP());
+  delay(500);
+
+  // Initialize Matter Node
+  Matter.begin();
+  // Initialize Matter EndPoint
+  OnOffLight.begin();
+  // Matter start
```
Comment (SuGlider) [inline libraries/Matter/examples/MatterCommissionTest/MatterCommissionTest.ino:45]:
```suggestion
  // Matter start - Last step, after all EndPoints are initialized

```

File: libraries/Matter/examples/MatterCommissionTest/MatterCommissionTest.ino [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+  Serial.println("====> Decommissioning in 30 seconds. <====");
+  delay(30000);
+  Matter.decommission();
+  Serial.println("Matter Node is decommissioned. Commsssioning widget shall start over.");
+}
+
+#else
+void setup() {}
+void loop() {}
+#endif /* CONFIG_ESP_MATTER_ENABLE_DATA_MODEL */
```
Comment (lucasssvaz) [inline libraries/Matter/examples/MatterCommissionTest/MatterCommissionTest.ino:74]: I don't think this is necessary as the macro is already required in the JSON to compile the sketch

File: libraries/Matter/examples/MatterComposedLights/MatterComposedLights.ino [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+  Serial.printf("Matter Light #1 is %s\r\n", OnOffLight1.getOnOff() ? "ON" : "OFF");
+  Serial.printf("Matter Light #2 is %s\r\n", OnOffLight2.getOnOff() ? "ON" : "OFF");
+  Serial.printf("Matter Light #3 is %s\r\n", OnOffLight3.getOnOff() ? "ON" : "OFF");
+  delay(3000);
+}
+
+#else
+void setup() {}
+void loop() {}
+#endif /* CONFIG_ESP_MATTER_ENABLE_DATA_MODEL */
```
Comment (lucasssvaz) [inline libraries/Matter/examples/MatterComposedLights/MatterComposedLights.ino:103]: I don't think this is necessary as the macro is already required in the JSON to compile the sketch

File: libraries/Matter/examples/MatterOnOffLight/MatterOnOffLight.ino [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+    Serial.println("Toggling Light!");
+    lastMillis = millis();
+    OnOffLight.toggle();  // Matter Controller also can see the change
+  }
+}
+
+#else
+void setup() {}
+void loop() {}
+#endif /* CONFIG_ESP_MATTER_ENABLE_DATA_MODEL */
```
Comment (lucasssvaz) [inline libraries/Matter/examples/MatterOnOffLight/MatterOnOffLight.ino:110]: I don't think this is necessary as the macro is already required in the JSON to compile the sketch

File: libraries/Matter/examples/MatterComposedLights/MatterComposedLights.ino (hunk 6 above)
Comment (SuGlider) [reply libraries/Matter/examples/MatterComposedLights/MatterComposedLights.ino:103]: If I remove it, it fails with ESP32S3 when CI compiles the examples.

Comment (lucasssvaz) [reply libraries/Matter/examples/MatterComposedLights/MatterComposedLights.ino:103]: But does S3 have support for Matter ? If yes, then something is wrong somewhere.

Comment (SuGlider) [reply libraries/Matter/examples/MatterComposedLights/MatterComposedLights.ino:103]: Yes, this is Matter over WiFi. It is suppose to work with ESP32, S2, S3, C3 and C6.
CI seems to pick some `sdkconfig.h` that has no Matter configuration, different from the current `release/v3.1.x` file that I get when I run `get.py` in my computer.

Comment (SuGlider) [reply libraries/Matter/examples/MatterComposedLights/MatterComposedLights.ino:103]: The commit that are right after this comment https://github.com/espressif/arduino-esp32/pull/10467#issuecomment-2415411880 are about trying to overcome CI errors.

I have tried many different possibilities. None of them have worked. 
The only way to get it to pass CI was using the current code.

Comment (lucasssvaz) [reply libraries/Matter/examples/MatterComposedLights/MatterComposedLights.ino:103]: If the sdkconfig didn't include `CONFIG_ESP_MATTER_ENABLE_DATA_MODEL` it wouldn't even try to compile the sketch for the chip. What was the issue exactly ? I will fix any CI related issues.

Comment (lucasssvaz) [reply libraries/Matter/examples/MatterComposedLights/MatterComposedLights.ino:103]: `#define CONFIG_ESP_MATTER_ENABLE_DATA_MODEL 1` only exists for QIO. The CI uses DIO by default. Overwrite the FQBNs and it should work properly.

Comment (lucasssvaz) [reply libraries/Matter/examples/MatterComposedLights/MatterComposedLights.ino:103]: Maybe we can move to using QIO by default as, with the new JSON options, it's easy to build in DIO if needed. I'll open a PR for that

Comment (SuGlider) [reply libraries/Matter/examples/MatterComposedLights/MatterComposedLights.ino:103]: Why DIO/QIO (flash mode) has to do with which `sdkconfig` is used? Is it something specific from CI? Please tell more about it.

Comment (SuGlider) [reply libraries/Matter/examples/MatterComposedLights/MatterComposedLights.ino:103]: Please check these 2 commit:
5960abede4c31e3ba2da79adf27dadafa2b468ab
9e74283a90cdc528e2e82efee091737c1be75bc7

Is it what you are talking about?

Comment (SuGlider) [reply libraries/Matter/examples/MatterComposedLights/MatterComposedLights.ino:103]: Overwriting CI FBQN for S3 with:

``` json
{
    "fqbn": {
      "esp32s3": [
        "espressif:esp32:esp32s3:PartitionScheme=huge_app,FlashMode=qio"
      ]
    },
    "requires": [
      "CONFIG_SOC_WIFI_SUPPORTED=y",
      "CONFIG_ESP_MATTER_ENABLE_DATA_MODEL=y"
    ]
}
```

Comment (lucasssvaz) [reply libraries/Matter/examples/MatterComposedLights/MatterComposedLights.ino:103]: > Why DIO/QIO (flash mode) has to do with which `sdkconfig` is used? Is it something specific from CI? Please tell more about it.

I also don't know, but the libs are only being built for QIO as you can see:

image

Comment (SuGlider) [reply libraries/Matter/examples/MatterComposedLights/MatterComposedLights.ino:103]: It was a problem with Lib Builder. It is fixed now and the new lib files shall be available by tomorrow.
I'll remove all unnecessary guards and ci.json content.

**Comment by SuGlider** [comment]
@lucasssvaz - All CI issues are solved with the new set of Lib Builder libraries. Changes done to examples were reverted.

**Comment by SuGlider** [comment]
@me-no-dev - The changes you requested are done. Including adding a Light Toggle Switch Button to the MatterOnOffLight.ino example.

This what I would like to merge by now. Please let me know.

**Comment by SuGlider** [comment]
@me-no-dev @lucasssvaz - I'll change a few other things and change the status to "In Review" by ebd of this week.

File: libraries/Matter/src/Matter.cpp [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+
+// This callback is called for every attribute update. The callback implementation shall
+// handle the desired attributes and return an appropriate error code. If the attribute
+// is not of your interest, please do not return an error code and strictly return ESP_OK.
+static esp_err_t app_attribute_update_cb(
+  attribute::callback_type_t type, uint16_t endpoint_id, uint32_t cluster_id,
+  uint32_t attribute_id, esp_matter_attr_val_t *val, void *priv_data)
+{
+  esp_err_t err = ESP_OK;
+  if (type == PRE_UPDATE) { /** Callback before updating the value in the database */
```
Comment (me-no-dev) [inline libraries/Matter/src/Matter.cpp:32]: Shouldn't this be a `switch` or `if/else if`?

Comment (SuGlider) [reply libraries/Matter/src/Matter.cpp:32]: It could be a `switch{}`.
I have just added the other "cases" for future use.
I'll change it.

Comment (SuGlider) [reply libraries/Matter/src/Matter.cpp]: Done.

File: libraries/Matter/src/Matter.cpp [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+// is not of your interest, please do not return an error code and strictly return ESP_OK.
+static esp_err_t app_attribute_update_cb(
+  attribute::callback_type_t type, uint16_t endpoint_id, uint32_t cluster_id,
+  uint32_t attribute_id, esp_matter_attr_val_t *val, void *priv_data)
+{
+  esp_err_t err = ESP_OK;
+  switch(type) {
+    case PRE_UPDATE: // Callback before updating the value in the database
+      log_i("Attribute update callback: PRE_UPDATE");
+      MatterEndPoint *ep = (MatterEndPoint *) priv_data;  // endpoint pointer to base class
```
Comment (me-no-dev) [inline libraries/Matter/src/Matter.cpp:35]: `MatterEndPoint *ep = NULL;` must be outside the switch or the whole case needs to be put in braces.

Comment (SuGlider) [reply libraries/Matter/src/Matter.cpp]: Solved

**Comment by SuGlider** [comment]
@me-no-dev - last change to example done. It keeps last Light state for when the Matter Accessory is restarted/powered back on.
All set and ready.

**Review by me-no-dev — APPROVED** [review]

**Review by lucasssvaz — APPROVED** [review]
LGTM

**Comment by me-no-dev** [comment]
@SuGlider please fix the issues. Matter does not yet run on P4

**Comment by SuGlider** [comment]
@me-no-dev - all set. Ready to merge

**Pull request merged**